apps/yapms/src/lib/types/Region.ts

```diff
@@ -24,6 +24,12 @@ export const RegionSchema = z.object({
 	})
 });
 
+export const RegionCandidateSchema = z.object({
+	candidate: CandidateSchema,
+	count: z.number(),
+	margin: z.number()
+});
+
```

Comment:
The "SavedRegionCandidatesSchema" could be based on this.

Then the RegionSchema could use the "SavedRegionCandidatesSchema".

Because this is duplicated 3 time otherwise.

Reply:
I think this should get a broader refactor where we standardize the ways in which region info is loaded. We have 4(?) different ways it is loaded in right now. I can make them derived from each other with the following code, but it is still messy.

```
import { CandidateSchema } from './Candidate';
import { z } from 'zod';

export const RegionCandidateSchema = z.object({
	candidate: CandidateSchema,
	count: z.number().nonnegative(),
	margin: z.number().nonnegative()
});

export const RegionSchema = z.object({
	id: z.string(),
	shortName: z.string(),
	longName: z.string(),
	value: z.number(),
	permaVal: z.number(),
	disabled: z.boolean(),
	locked: z.boolean(),
	permaLocked: z.boolean(),
	candidates: RegionCandidateSchema.array(),
	nodes: z.object({
		region: z.custom<HTMLElement>((val) => val instanceof HTMLElement),
		button: z.custom<HTMLElement>((val) => val instanceof HTMLElement).nullable(),
		text: z.custom<HTMLElement>((val) => val instanceof HTMLElement).nullable()
	})
});

export const SavedRegionCandidateSchema = RegionCandidateSchema.omit({ candidate: true }).extend({
	candidate:z.string()
})

export const SavedRegionCandidatesSchema = SavedRegionCandidateSchema.array()

export const SavedRegionSchema = RegionSchema.omit({
	shortName: true,
	longName: true,
	nodes: true,
	candidates: true
}).extend({
	candidates: SavedRegionCandidateSchema.omit({ candidate: true }).extend({
		id:z.string()
	}).array()
});

type Region = z.infer<typeof RegionSchema>;

export type RegionCandidate = z.infer<typeof RegionCandidateSchema>;

export type SavedRegionCandidates = z.infer<typeof SavedRegionCandidatesSchema>;

export default Region;
```

Reply:
Oh sorry gottcha, I wasn't paying attention to the fact that the SavedRegionCandidatesSchema represented the ID of the candidate in a file.

If you want you could just do this to remove the duplicates. Otherwise it's fine.

```ts
export const RegionCandidateSchema = z.object({
	candidate: CandidateSchema,
	count: z.number(),
	margin: z.number()
});

export const RegionSchema = z.object({
	id: z.string(),
	shortName: z.string(),
	longName: z.string(),
	value: z.number(),
	permaVal: z.number(),
	disabled: z.boolean(),
	locked: z.boolean(),
	permaLocked: z.boolean(),
	candidates: RegionCandidateSchema.array(),
	nodes: z.object({
		region: z.custom<HTMLElement>((val) => val instanceof HTMLElement),
		button: z.custom<HTMLElement>((val) => val instanceof HTMLElement).nullable(),
		text: z.custom<HTMLElement>((val) => val instanceof HTMLElement).nullable()
	})
});

export const SavedRegionCandidatesSchema = z
	.object({
		candidate: z.string(),
		count: z.number().nonnegative(),
		margin: z.number().nonnegative()
	})
	.array();

export const SavedRegionSchema = RegionSchema.omit({
	shortName: true,
	longName: true,
	nodes: true,
	candidates: true
}).extend({
	candidates: SavedRegionCandidatesSchema
});
```

Reply:
This breaks loading maps so I will leave it out and we can refactor at another time.